src/NodeVisitor/OptimizerNodeVisitor.php

```diff
@@ -107,13 +110,20 @@ private function mergeTextNodeCalls(Node $node): Node
             return $node;
         }
 
+        $firstChildNode = $node->getNode($names[0]);
+        $line = $firstChildNode->getTemplateLine();
+        $sourceContext = $firstChildNode->getSourceContext();
         if (Node::class === get_class($node)) {
-            return new TextNode($text, $node->getTemplateLine());
+            $textNode = new TextNode($text, $line);
+            $textNode->setSourceContext($sourceContext);
```

Comment:
this will break if `$sourceContext` is null.

But maybe the right fix would be for `\Twig\Node\Node::setNode` to set the source context in the new node in case we have a source context and the node being set does not have one. This would be consistent with the fact that `setSourceContext` also sets the node on all children at the time it is set.
this way, it would work for the nodes created by OptimizerNodeVisitor but would also solve the similar issue for any other custom node visitor

Reply:
Setting the `SourceContext` in `setNode` sounds like a great idea. That solves my current problem with missing `SourceContexts` after optimization and simplifies all of my current `NodeVisitor` use cases where i have to set the `SourceContext` everywhere i create new nodes.

I created a separate PR for that small change: https://github.com/twigphp/Twig/pull/4050